pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,303 @@
+	// that the request accesses.  If the object is not namespaced or the request is a non-resource request then
+	// the flow distinguisher is the empty string.
+	FlowDistinguisherMethodByNamespaceType FlowDistinguisherMethodType = "ByNamespace"
+
+	GroupKind          = "Group"
+	ServiceAccountKind = "ServiceAccount"
+	UserKind           = "User"
+
+	APIGroupAll    = "*"
+	ResourceAll    = "*"
+	VerbAll        = "*"
+	NonResourceAll = "*"
+	NameAll        = "*"
+)
+
+// System preset priority level names
+const (
+	PriorityLevelConfigurationNameExempt = "exempt"
+)
+
+// Default settings for flow-schema
+const (
+	FlowSchemaDefaultMatchingPrecedence int32 = 1000
+)
+
+// Default settings for priority-level-configuration
+const (
+	PriorityLevelConfigurationDefaultHandSize int32 = 1
+)
+
+// Conditions
+const (
+	FlowSchemaConditionDangling = "Dangling"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow is made up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// `spec` is the specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// `status` is the current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// `metadata` is the standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// `items` is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// `priorityLevelConfiguration` should reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will be ignored and marked as invalid in its status.
+	// Required.
+	PriorityLevelConfiguration PriorityLevelConfigurationReference
+	// `matchingPrecedence` is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
```

Comment:
Please be explicit about zero.  It looks positive for me.

Reply:
I am not sure I understand.  This field is given a specific range of allowed values, and their only meaning is their relative values.  I am not sure what more needs to be said.  I do not understand what you mean by "It looks positive for me".

Reply:
> 	// Note that if the precedence is not specified or zero, it will be set to 1000 as default.


i added this to the comment, wdyt?

Reply:
> > // Note that if the precedence is not specified or zero, it will be set to 1000 as default.
> 
> i added this to the comment, wdyt?

perfect.  That's exactly what we needed.

Reply:
Oh, that.
I think it is a mistake to have a default value for this field.  The rules need to be in a definite, designed, and understood order.  It is wrong to suggest not thinking about the numbers that give the ordering.

Reply:
i'm slightly leaning on making the precedence default'able:

(1) we already have a long list of fields required when creating a new FS, it's challenging user-experience when we're forcing users to think about every fields' meaning..
(2) the field can always be updated, defaulting a low precedence is "safe", users can promote it if they want afterwards.
(3) passive conflict resolving. if the field is required, users will have to resolve the overlapping w/ other FS's in the first place. otherwise, users will only have to resolve it when they found conflict between rules. consider 100-ish FS in a cluster, this will also lead to better user-experience.

Reply:
Having two flow schemas with the same matching precedence creates an ambiguity if there is any request that both schemas match.  This is going to force a fair amount of different settings of matching precedence all by itself.

Worse yet, "optimizing" by using the same matching precedence for flow schemas that do _not_ overlap (in terms of the set of matched requests) creates a time-bomb.  Consider what happens later when an admin wants to add a third flow schema that overlaps with one of the first two, and should be ordered between them.  There is no matching precedence value that can be assigned to the new flow schema that accomplishes this.

We do not have 100-ish flow schemas in a cluster.  The KEP's exhibited configuration has five.  It is always easy to ask "what if this number gets big", but without a use case for it we are making our problem harder than it has to be.

If a cluster _did_ have 100-ish flow schemas, it will be easier to understand how they interact if they are in a definite order.